### PR TITLE
Make iface-error-test-1 less flaky

### LINKED ISSUE #2879: ghcide test "iface-error-test-1" is flaky
The test has at least 3 failure modes, here is a [test run](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true) showing all three:

## Failure 1: timeout
```
Interface loading tests
[964](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:964)
    iface-error-test-1:                                                                                   FAIL
[965](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:965)
      Exception: Timed out waiting to receive a message from the server.
[966](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:966)
      Last message received:
[967](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:967)
      {
[968](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:968)
          "method": "window/logMessage",
[969](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:969)
          "params": {
[970](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:970)
              "message": "Live bytes: 0.00MB Heap size: 0.00MB",
[971](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:971)
              "type": 3
[972](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:972)
          },
[973](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:973)
          "jsonrpc": "2.0"
[974](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:974)
      }
[975](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:975)
      Use -p '/iface-error-test-1/' to rerun this test only.
```
## Failure 2: unexpected diagnostics for `P.hs`:
```
Interface loading tests
[1042](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:1042)
    iface-error-test-1: FAIL (2.50s)
[1043](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:1043)
      test/exe/Main.hs:6107:
[1044](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:1044)
      Got unexpected diagnostics for Uri {getUri = "file:///tmp/extra-dir-11038963127873/P.hs"} got [Diagnostic {_range = Range {_start = Position {_line = 4, _character = 0}, _end = Position {_line = 4, _character = 3}}, _severity = Just DsWarning, _code = Just (InR "-Wmissing-signatures"), _source = Just "typecheck", _message = "Top-level binding with no type signature: bar :: Int", _tags = Nothing, _relatedInformation = Nothing},Diagnostic {_range = Range {_start = Position {_line = 6, _character = 0}, _end = Position {_line = 6, _character = 3}}, _severity = Just DsWarning, _code = Just (InR "-Wmissing-signatures"), _source = Just "typecheck", _message = "Top-level binding with no type signature: foo :: Bool", _tags = Nothing, _relatedInformation = Nothing},Diagnostic {_range = Range {_start = Position {_line = 4, _character = 0}, _end = Position {_line = 4, _character = 3}}, _severity = Just DsWarning, _code = Just (InR "-Wmissing-signatures"), _source = Just "typecheck", _message = "Top-level binding with no type signature: bar :: Int", _tags = Nothing, _relatedInformation = Nothing},Diagnostic {_range = Range {_start = Position {_line = 6, _character = 0}, _end = Position {_line = 6, _character = 3}}, _severity = Just DsWarning, _code = Just (InR "-Wmissing-signatures"), _source = Just "typecheck", _message = "Top-level binding with no type signature: foo :: Bool", _tags = Nothing, _relatedInformation = Nothing}]
```
## Failure 3: unexpected diagnostics for `A.hs`:
```
FAIL (2.54s)
[3100](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:3100)
      test/exe/Main.hs:6107:
[3101](https://github.com/haskell/haskell-language-server/runs/6245281626?check_suite_focus=true#step:8:3101)
      Got unexpected diagnostics for Uri {getUri = "file:///tmp/extra-dir-15590457777873/A.hs"} got []
```

### PR DESCRIPTION
**Pull request by pepeiborra**
Fixes #2879 by ensuring that the barrage of graph builds triggered by the test are observed as deterministically as possible, which amounts to inserting calls to `waitForProgressDone` and using `openDoc` instead of `createDoc`. 

While I was there, I also cleaned up a bunch of minor things in the ghcide codebase

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2882"></a>

### CODE REVIEW

File: ghcide/test/src/Development/IDE/Test.hs [hunk 1]
```diff
@@ -83,7 +83,7 @@ expectNoMoreDiagnostics timeout =
   expectMessages STextDocumentPublishDiagnostics timeout $ \diagsNot -> do
     let fileUri = diagsNot ^. params . uri
         actual = diagsNot ^. params . diagnostics
-    liftIO $
+    unless (actual == List []) $ liftIO $
```
Comment (michaelpj) [inline ghcide/test/src/Development/IDE/Test.hs:86]: Huh, does this actually happen?

File: ghcide/test/exe/Main.hs [hunk 2]
```diff
@@ -6075,11 +6075,14 @@ ifaceErrorTest = testCase "iface-error-test-1" $ runWithExtraFiles "recomp" $ \d
     expectDiagnostics
       [("P.hs", [(DsWarning,(4,0), "Top-level binding")])] -- So what we know P has been loaded
 
+    waitForProgressDone
```
Comment (michaelpj) [inline ghcide/test/exe/Main.hs:6078]: Worth some comments explaining that these calls allows the changes to settle?

**Review by michaelpj — APPROVED** [review]

File: ghcide/test/src/Development/IDE/Test.hs (hunk 1 above)
Comment (pepeiborra) [reply ghcide/test/src/Development/IDE/Test.hs:86]: Yes, this is actually a 4th failure vector for this test

File: ghcide/test/exe/Main.hs (hunk 2 above)
Comment (pepeiborra) [reply ghcide/test/exe/Main.hs:6078]: The 100s of existing unit tests implicitly document this approach. Even if I added a few comments here, they would not be discoverable enough to be helpful

**Pull request merged**